### PR TITLE
Added mute and unmute command

### LINKED ISSUE #6: [feature] Add mute command
`;mute <user> <time> <reason>`
user can be userid, username, mention
time can be 5m (5 minutes), 1h (1 hour), 3d (3 days)
reason is the rest of argv joined by space can be any length

This would apply the mute role to the user for the duration of time specified. Mute role is specified in config by key `muteRole` and mute must be added to a new Mutes table with serverId, userId, byUserId, duration, reason, creation

### PR DESCRIPTION
**Pull request by notfooddev**

### CODE REVIEW

File: databaseCreation.sql [hunk 1]
```diff
@@ -43,4 +45,15 @@ CREATE TABLE `WhiteListedRoles` (
...
+
+DROP TABLE IF EXISTS `Muted`;
+CREATE TABLE `Muted` (
+  `guildId` VARCHAR(255) NOT NULL,
+  `userId` varchar(255) NOT NULL,
+  `byUserId` varchar(255) DEFAULT NULL,
+  `date` datetime DEFAULT NULL,
+  `until` datetime DEFAULT NULL,
+  `reason` longtext,
+  PRIMARY KEY (`guildId`, `userId`)
```
Comment (WesVleuten) [inline databaseCreation.sql]: This will throw an error if a user has been muted a second time

**Review by WesVleuten — CHANGES_REQUESTED** [review]
Haven't yet looked at everything but this is blocking further testing

File: databaseCreation.sql (hunk 1 above)
Comment (notfooddev) [reply databaseCreation.sql]: It shouldn't as every time someone it muted, it checks if this person is already muted and removes this entry from the database for the new mute. There was however a separate little bug in this, so it's better to have checked anyway.

File: src/commands/config.ts [hunk 2]
```diff
@@ -193,6 +214,39 @@ export default class ConfigCommand implements ICommand {
...
+					const muteRole = guild.roles.cache.find(x => x.id == value || x.name == value);
+					if (!muteRole) {
+						message.reply('Couldnt find role');
+						return;
+					}
+					ss.muteRole = muteRole.id;
+				}
+			} else if (key == 'muteChannel') {
+				if (value == 'null') {
+					ss.muteChannel = null;
```
Comment (WesVleuten) [inline src/commands/config.ts:230]: When a mute channel is unset, the permissions are not restored

File: src/commands/mute.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+			return;
+		}
+
+		if (args.length < 3 || message.mentions.members === null || message.mentions.members.size === 0) { 
+			return;
+		}
+
+		const guildMember = message.mentions.members.first()
+		if (!guildMember || !args[0].includes(guildMember.id) || guildMember.id === message.author.id) { 
+			return;
```
Comment (WesVleuten) [inline src/commands/mute.ts:40]: "User not found"
"You cannot mute yourself"

File: src/commands/mute.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	
+	async run(discordClient: Client, message: Message, args: string[]) {
+		const guildId = message.guild?.id;
+		const serverSettings = await ServerSettingsRepository.GetByGuildId(guildId);
+		if (serverSettings === null || serverSettings.muteRole === null) {
+			return;
+		}
+
+		if (args.length < 3 || message.mentions.members === null || message.mentions.members.size === 0) { 
+			return;
```
Comment (WesVleuten) [inline src/commands/mute.ts:29]: Send help text

File: src/commands/mute.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	guildOnly = false;
+
+	usageText = ";mute <user> <time> <reason>";
+	helpText = "Mutes user";
+	
+	async run(discordClient: Client, message: Message, args: string[]) {
+		const guildId = message.guild?.id;
+		const serverSettings = await ServerSettingsRepository.GetByGuildId(guildId);
+		if (serverSettings === null || serverSettings.muteRole === null) {
+			return;
```
Comment (WesVleuten) [inline src/commands/mute.ts:24]: Should display something like "Muting has not been configured"

File: src/commands/mute.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+					key: "Reason",
+					value: `${reason}`,
+					inline: false,
+				}
+			],
+		})
+		
+		const muteRole = await message.guild?.roles.fetch(serverSettings.muteRole);
+		if (!muteRole) {
+			return;
```
Comment (WesVleuten) [inline src/commands/mute.ts:69]: Shoud return an error message to the user

File: src/commands/unmute.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	aliases = null;
+	permissionLevel = PermissionLevel.Moderator;
+	guildOnly = false;
+
+	usageText = ";unmute <user>";
+	helpText = "Unmutes user";
+	
+	async run(discordClient: Client, message: Message, args: string[]) {
+		if (args.length !== 1 || message.mentions.members === null || message.mentions.members.size === 0) { 
+			return;
```
Comment (WesVleuten) [inline src/commands/unmute.ts:19]: Should return command usage

File: src/commands/unmute.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		}
+
+		const guildId = message.guild?.id;
+		const serverSettings = await ServerSettingsRepository.GetByGuildId(guildId);
+		if (serverSettings === null || serverSettings.muteRole === null) {
+			return;
```
Comment (WesVleuten) [inline src/commands/unmute.ts:26]: Should return "misconfigured" response

File: src/commands/unmute.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+		}
+
+		const muteRole = await message.guild?.roles.fetch(serverSettings.muteRole);
+		if (!muteRole) {
+			return;
+		}
+
+		const guildMember = message.mentions.members.first()
+		if (!guildMember || !args[0].includes(guildMember!.id)) { 
+			return;
```
Comment (WesVleuten) [inline src/commands/unmute.ts:38]: Should return user not found error

File: src/commands/unmute.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+		const mute = await MutedRepository.GetRunning(guildId, guildMember.id)
+		if (mute === null) { 
+			return;
+		}
+
+		MutedRepository.SetUnmuted(mute.id, new Date())
+		const user = message.guild!.members.resolve(guildMember.id);
+		if (user === null) { 
+			return;
```
Comment (WesVleuten) [inline src/commands/unmute.ts:49]: Should return user not found error

File: src/commands/unmute.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		}
+
+		const guildMember = message.mentions.members.first()
+		if (!guildMember || !args[0].includes(guildMember!.id)) { 
+			return;
+		}
+
+		const mute = await MutedRepository.GetRunning(guildId, guildMember.id)
+		if (mute === null) { 
+			return;
```
Comment (WesVleuten) [inline src/commands/unmute.ts:43]: Should return (maybe) something like user is not muted

File: src/events/channelCreate.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+import ServerSettingsRepository from "../repository/serverSettings";
+import { SetMutedPermissionsForChannel } from "../lib/mutedRole";
+
+export default async function ChannelCreateEvent(discordClient: DiscordClient, channel: GuildChannel) {
+	let serverSettings = await ServerSettingsRepository.GetByGuildId(channel.guild.id)
+	if (serverSettings === null) {
+		return;
+	}
+
+	if (serverSettings!.muteRole !== null) {
```
Comment (WesVleuten) [inline src/events/channelCreate.ts]: Remove the `serverSettings === null` check and place `serverSettings?.muteRole != null` here

File: src/events/ready.ts [hunk 13]
```diff
@@ -1,18 +1,35 @@
...
+		let serverSettings = await ServerSettingsRepository.GetByGuildId(guild.id)
 		if (serverSettings === null) {
-			ServerSettingsRepository.Create(guildId);
-			Logger.info(`Created guild ${guildId}`);
+			ServerSettingsRepository.Create(guild.id);
+			Logger.info(`Created guild ${guild.id}`);
+		}
+		
+		serverSettings = await ServerSettingsRepository.GetByGuildId(guild.id)
+		if (serverSettings!.muteRole !== null) {
```
Comment (WesVleuten) [inline src/events/ready.ts]: Better to user a ? here instead of !

**Review by WesVleuten — CHANGES_REQUESTED** [review]
Using `!` is very dangerous. This could introduce null reference errors. Please use `?` if possibble

File: src/commands/unmute.ts (hunk 10 above)
Comment (notfooddev) [reply src/commands/unmute.ts:49]: I think this doesn't really matter

File: src/commands/unmute.ts (hunk 11 above)
Comment (notfooddev) [reply src/commands/unmute.ts:43]: Feels unnecessary to me

File: src/commands/config.ts [hunk 14]
```diff
@@ -193,6 +213,54 @@ export default class ConfigCommand implements ICommand {
...
+					const muteChannel = guild.channels.cache.find(x => x.id == value || x.name == value);
```
Comment (WesVleuten) [inline src/commands/config.ts]: `value` here is `'null'`, does this even work?

File: src/commands/mute.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+			message.reply(this.usageText)
+			return;
+		}
+		
+		if (guildMember.id === message.author.id) { 
+			message.reply("User can't mute themselves");
+			return;
+		}
+
+		const date = this.parseDate(args[1]);
```
Comment (WesVleuten) [inline src/commands/mute.ts:43]: Muting for a longer time, like a month, results in a console error and the mute not working.

```
(node:926) TimeoutOverflowWarning: 2678400136 does not fit into a 32-bit signed integer.
Timeout duration was set to 1.
```

**Review by WesVleuten — CHANGES_REQUESTED** [review]
Looking good, almost there

**Review by WesVleuten — COMMENTED** [review]
Looks good, but still unsure if we should push it through without the channel overwrite role support

File: src/events/channelCreate.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import { Client as DiscordClient, GuildChannel } from "discord.js";
+import ServerSettingsRepository from "../repository/serverSettings";
+import { SetMutedPermissionsForChannel } from "../lib/mutedRole";
+
+export default async function ChannelCreateEvent(discordClient: DiscordClient, channel: GuildChannel) {
```
Comment (WesVleuten) [inline src/events/channelCreate.ts:5]: This event does not get called on channelCreate.

Please see `EventBind` in `routes.ts`

**Review by WesVleuten — CHANGES_REQUESTED** [review]

File: src/events/channelCreate.ts (hunk 16 above)
Comment (notfooddev) [reply src/events/channelCreate.ts:5]: Weird, this was explicitly tested. I guess it got lost somewhere on the way.

**Review by WesVleuten — APPROVED** [review]
Hell yeah

**Pull request merged**